#### $\color{gray}{\text{Autor: Maciej S.}}$

# $\textcolor{teal}{\textbf{Schemat Bernoulliego}}$
$\textcolor{teal}{\textbf{Próbą Bernoulliego}}$ nazywamy doświadczenie, w którym otrzymujemy jeden z dwóch możliwych wyników. Jeden z tych wyników nazywamy $\textcolor{olive}{\text{sukcesem}}$, a drugi $\textcolor{olive}{\text{porażką}}$. Jeżeli prawdopodobieństwo sukcesu wynosi $\textcolor{olive}{p}$, to prawdopodobieństwo porażki $\textcolor{olive}{q = 1 - p}$.
$\newline$
$\textcolor{teal}{\textbf{Schematem Bernoulliego}}$ nazywamy ciąg niezależnych powtórzeń prób Bernoulliego.
$\newline$
W $\textcolor{teal}{\textbf{schemacie Bernoulliego}}$ prawdopodobieństwo uzyskania $\textcolor{olive}{\text{k}}$ sukcesów w $\textcolor{olive}{\text{n}}$ próbach mozna obliczyć ze wzoru:
$\newline$
$$ \Large {P_n(k) = \binom{n}{k} \cdot p^k \cdot q^{n-k}} $$
$\newline$
### Przykład: 
$\newline$ Rzucamy sześcienną kostką do gry 4 razy. Jakie jest pradopodobieństwo, że pięć oczek wypadnie dokładnie dwa razy?
######
Liczba prób: $\textcolor{olive}{n = 4}$ $\newline$ Liczba oczekiwanych sukcesów: $\textcolor{olive}{k = 2}$ $\newline$ Prawdopodobieńswto, że wypdanie pięć oczek w pojedynczym rzucie: $ \textcolor{olive}{{p = \frac{1}{6}}}$ $\newline$ Prawdopodobieńswto, że wypadnie wypadnie inna liczba oczek: $\textcolor{olive}{q = p - 1 = \frac{5}{6}}$ 
#
$$ P_4(2) = \binom{4}{2} \cdot \left( \frac{1}{6} \right) ^{2} \cdot \left( \frac{5}{6} \right)^{4-2} = \frac{4!}{(4-2)! \cdot 2!} \cdot \frac{1}{36} \cdot \frac{25}{36} = \frac{24}{2 \cdot 2} \cdot \frac{1}{36} \cdot \frac{25}{36} = \frac{600}{5184} = \frac{25}{216} \approx 0.115740741 $$
##### Poniżej możesz sprawdzić prawdopodobieństwo wypadnięcia konkretnej liczby oczek, kokretną liczbę razy, dla danej liczby prób.

In [2]:
from fractions import Fraction
import math
n = int(input("Ile razy chcesz rzucić: "))
k = int(input("Ile razy ma wypaść twoja liczba: "))
p = Fraction(1, 6)
q = Fraction(5, 6)
x = Fraction(math.factorial(n), math.factorial(n - k) * math.factorial(k))
odp_frac = Fraction(x * p**k * q**(n-k), 1)
print("Prawdopodobieństwo tego zdarzenia to ", odp_frac, " co w przybliżeniu jest równe: ", float(odp_frac), ".")

Prawdopodobieństwo tego zdarzenia to  25/216  co w przybliżeniu jest równe:  0.11574074074074074 .


# $\textcolor{teal}{\textbf{Rozkład Poissona}}$
Niech zmienna losowa $ \color{olive}X_n $ ma rozkład Bernoulliego określony wzorem:
$$ \large {P(X_n = k) = \binom{n}{k} \cdot p^k \cdot q^{n-k}}, \quad k = 0, 1, 2, ..., n $$
Załóżmy, że liczba $\color{olive}{n}$ dąży do nieskończoności i iloczyn $\color{olive}{n \cdot p}$ jest stały, tzn. $\color{olive}{n \cdot p = \lambda}$, $\color{olive}{\lambda}$ - stała dodatnia. Tak określona nowa zmienna losowa $\color{olive}{X}$ może przyjąć każdą wartość całkowitą z pradziału $\color{olive}{ \langle 0, \infty \rangle }$. Prawdopodobieństwo przyjęcia przez tę zmienną wartości $\color{olive}{k}$ wyraża się $\color{teal}{\text{wzorem Poissona}}$, tj.
$$ \large P(X = k) = \lim_{x\to\infty}P(X_n = k) = \frac{e^{-\lambda} \cdot \lambda ^k}{k!}, \quad k = 0, 1, 2, ...  $$
### $\textbf{Definicja}$
Zmienna losowa $ X $ ma $\color{teal}{\text{rozkład Poissona}}$ z parametrem $\color{olive}{\lambda = n \cdot p} $, jesli jej funkcja prawdopodobieństwa dana jest wzorem:
$$ \large P(X = k) = \frac{e^{-\lambda} \cdot \lambda ^k}{k!}, \quad \lambda > 0, \quad k = 0, 1, 2, ...  $$
Mówimy, że $\color{teal}{\text{rozkład Poissona}}$ jest rozkładem granicznym $ \color{teal}{\text{rozkładu Bernoulliego}} $ przy podanych warunkach lub że $ \color{teal}{\text{rozkład Bernoulliego}} $ jest zbieżny do $ \color{teal}{\text{rozkładu Poissona}} $. $\newline$
##### Poniżej program porównujący pradopodobieństwa obliczone dwiema metodami.

In [23]:
n = int(input("Liczba prób: "))
k = int(input("Liczba sukcesów: "))
s = int(input("Liczba możliwych wyborów (zakładamy, że sukces można osiągnąć w jeden sposób): "))
p = Fraction(1, s)
q = Fraction(s - 1, s)
x = Fraction(math.factorial(n), math.factorial(n - k) * math.factorial(k))
odp_frac_bernoullie = Fraction(x * p**k * q**(n-k), 1)
odp_frac_poisson = Fraction(int(math.e**(-n*p) * (n*p)**k), math.factorial(k))
print("Prawdopodobieństwo tego zdarzenia wyliczone za pomocą schematu Bernoulliego w przybliżeniu jest równe:", float(odp_frac_bernoullie), ".")
print("Prawdopodobieństwo tego zdarzenia wyliczone za pomocą rozkładu Poissona w przybliżeniu jest równe:", float(odp_frac_poisson), ".")
print("Różnica pomiędzy prawdopodobieństwami to: ", float(odp_frac_bernoullie - odp_frac_poisson))

Prawdopodobieństwo tego zdarzenia wyliczone za pomocą schematu Bernoulliego w przybliżeniu jest równe: 0.0892828157326504 .
Prawdopodobieństwo tego zdarzenia wyliczone za pomocą rozkładu Poissona w przybliżeniu jest równe: 0.0888353173920853 .
Różnica pomiędzy prawdopodobieństwami to:  0.00044749834056509945


### $\color{teal}{\textbf{Porównanie wyników}}$
Tabela porównująca prawdopodobieństwa dla zajścia $\color{olive}{\text{k}}$ sukcesów po $\color{olive}{\text{n}}$ próbach, gdzie pradopodobieństwo pojedynczego zdarzenia jest równe $\color{olive}{\text{p}}$. Dokładne, wyliczone $\color{teal}{\text{schematem Bernoulliego}}$, i przybliżone - wyliczone $\color{teal}{\text{rozkładem Poissona}}$.
##### Wyniki dla $\color{olive}{k = n \cdot p}$ :
| | wartość | $\color{olive}{\bf{n = 100}}$ | $\color{olive}{\bf{n = 200}}$ | $\color{olive}{\bf{n = 500}}$ | $\color{olive}{\bf{n = 1000}}$ | $\color{olive}{\bf{n = 2000}}$ |
| - | - | - | - | - | - | - |
| $\color{olive}{\bf{p = \frac{1}{20}}}$ | dokładna | 0.180 | 0.12836 | 0.08159 | 0.05779 | 0.04090 |
| $\color{olive}{\bf{p = \frac{1}{20}}}$ | przybliżona | 0.175 | 0.12511 | 0.07952 | 0.05633 | 0.03986|
| | $\textbf{różnica}$ | 0.005 | 0.00324 | 0.00207 | 0.00146 | 0.00103 |
| $\color{olive}{\bf{p = \frac{1}{50}}}$ | dokładna | 0.27341 | 0.19735 | 0.12638 | 0.08974 | 0.06359 |
| $\color{olive}{\bf{p = \frac{1}{50}}}$ | przybliżona | 0.0 | 0.16667 | 0.12511 | 0.08884 | 0.06295 |
| | $\textbf{różnica}$ | 0.27341 | 0.03068 | 0.00127 | 0.00090 | 0.00064 |
| $\color{olive}{\bf{p = \frac{1}{100}}}$ | dokładna | 0.36973 | 0.27203 | 0.17635 | 0.12574 | 0.08928 |
| $\color{olive}{\bf{p = \frac{1}{100}}}$ | przybliżona | 0.0 | 0.0 | 0.17500 | 0.12511 | 0.08884 |
| | $\textbf{różnica}$ | 0.36972 | 0.27203 | 0.00135 | 0.00063 | 0.00045 |
#### $\textbf{Wnioski:}$
Wraz ze wzrostem liczby $\textcolor{olive}{\text{prób}}$ oraz zmniejszaniem się prawdopodobieństwa pojedynczego $\textcolor{olive}{\text{zdarzenia}}$, $\textcolor{teal}{\text{schemat Poissona zbiega do schematu Bernoulliego}}$. 